### PR TITLE
osd: improved handling of new-interval events by the scrubber

### PR DESCRIPTION
**Pull request by ronen-fr**
Following @athanatos code-review comments to PR #38111, this PR changes the way the scrubber handles interval changes:

- the scrubber is now notified of new intervals directly (vs. waiting for the next scrub-related event in order to "notice" the change)
- remote reservations are specifically discarded
- we no longer expect to remain in Primary role after an interval change (fixing a reported bug)

Fixes: https://tracker.ceph.com/issues/48712
Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

**Comment by ronen-fr** [comment]
Note re testing:
The scrub test suit passes with the following bug-fix PRs:
#38726 
#38723

**Comment by ronen-fr** [comment]
partial testing (rados suite, "scrub", subset 3/5) OK
http://pulpito.front.sepia.ceph.com/rfriedma-2020-12-29_13:24:41-rados-wip-ronenf-scrub-interval-distro-basic-smithi

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-12-30_06:47:01-rados-wip-kefu-testing-2020-12-30-1123-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-31_04:10:08-rados-wip-kefu-testing-2020-12-30-1123-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-31_09:50:24-rados-wip-kefu-testing-2020-12-31-1427-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-01_04:13:56-rados-wip-kefu-testing-2020-12-31-1427-distro-basic-smithi/

**Comment by tchaikov** [comment]
> Note re testing:
> The scrub test suit passes with the following bug-fix PRs:
> #38726
> #38723

@ronen-fr could you please merge these two PRs into one so i can repeat the test? they conflict with each other when i try to test them in a single branch. but as you put, the scrub test passes with them.

**Comment by ronen-fr** [comment]
@tchaikov 
> could you please merge these two PRs into one so i can repeat the test? they conflict with each other when i try to test them in a single branch. but as you put, the scrub test passes with them.

WIll do today. Thanks

File: src/osd/osd_types.h [hunk 1]
```diff
@@ -1110,7 +1110,7 @@ class pool_opts_t {
   }
 
   template<typename T>
-  T value_or(key_t key, const T& default_value) const {
+  T value_or(key_t key, T& default_value) const {
```
Comment (athanatos) [inline src/osd/osd_types.h:1113]: This doesn't seem right, where was it selecting the wrong signature?

Comment (ronen-fr) [reply src/osd/osd_types.h:1113]: @athanatos,
I forgot that '&&' in template functions is "perfect forwarding / catch-all", and not "R-ref".
See https://godbolt.org/z/dMjf77 for what happens.

Now, digging further, I should probably remove this overload altogether.

Comment (athanatos) [reply src/osd/osd_types.h:1113]: Right, and you'll want to change std::move in the other overload to std::forward.

Comment (ronen-fr) [reply src/osd/osd_types.h:1113]: fixed

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-01-06_02:57:51-rados-wip-kefu-testing-2021-01-05-2058-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-07_03:01:15-rados-wip-kefu-testing-2021-01-05-2058-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-08_08:11:44-rados-wip-kefu-testing-2021-01-08-1210-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-10_03:55:58-rados-wip-kefu-testing-2021-01-08-1210-distro-basic-smithi/

**Comment by ronen-fr** [comment]
@tchaikov Thanks for testing. Going over the results now.

**Comment by neha-ojha** [comment]
@ronen-fr can you please verify if this PR is addressing all the three bugs mentioned in https://tracker.ceph.com/issues/48777#note-7?

**Comment by ronen-fr** [comment]
One of the tests (kchai-2021-01-06_02:57:51-rados-wip-kefu-testing-2021-01-05-2058-distro-basic-smithi/5758565)
displays a problem in the handling of a "blocked range" scenario. Will create a tracker shortly.

**Comment by ronen-fr** [comment]
> @ronen-fr can you please verify if this PR is addressing all the three bugs mentioned in https://tracker.ceph.com/issues/48777#note-7?

@neha-ojha 
- 48777, 48793 - seems so
- 48712 - definitely

(but note that there is a rarer scenario detected in @tchaikov tests, involving blocked ranges that are never release on time, that I need to handle).

**Comment by neha-ojha** [comment]
> > @ronen-fr can you please verify if this PR is addressing all the three bugs mentioned in https://tracker.ceph.com/issues/48777#note-7?
> 
> @neha-ojha
> 
> * 48777, 48793 - seems so

@ronen-fr I'm afraid not,

https://pulpito.ceph.com/kchai-2021-01-06_02:57:51-rados-wip-kefu-testing-2021-01-05-2058-distro-basic-smithi/5758213/ (among others) is failing with 48777

https://pulpito.ceph.com/kchai-2021-01-06_02:57:51-rados-wip-kefu-testing-2021-01-05-2058-distro-basic-smithi/5758415/ is failing with 48793

Also seeing https://pulpito.ceph.com/kchai-2021-01-06_02:57:51-rados-wip-kefu-testing-2021-01-05-2058-distro-basic-smithi/5758616/

> * 48712 - definitely
> 
> (but note that there is a rarer scenario detected in @tchaikov tests, involving blocked ranges that are never release on time, that I need to handle).

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-01-10_14:19:31-rados-wip-kefu2-testing-2021-01-10-1949-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-01-11_11:52:22-rados-wip-kefu2-testing-2021-01-10-1949-distro-basic-smithi/

**Comment by neha-ojha** [comment]
Being tested here: https://pulpito.ceph.com/nojha-2021-01-13_23:51:30-rados-wip-38730-distro-basic-smithi/

**Comment by neha-ojha** [comment]
last run (https://pulpito.ceph.com/nojha-2021-01-13_23:51:30-rados-wip-38730-distro-basic-smithi/) is not valid since the sha1 has changed, testing again

**Comment by neha-ojha** [comment]
This PR along with https://github.com/ceph/ceph/pull/38860 seems to fix the following issues that I had spotted
- https://tracker.ceph.com/issues/48777 (now deleted)
- https://tracker.ceph.com/issues/48793
- https://tracker.ceph.com/issues/48712

Not seeing related failures in https://pulpito.ceph.com/nojha-2021-01-15_01:03:38-rados-wip-38730-2021-01-14-distro-basic-smithi/

@athanatos can you take another look? I'd like to merge it to address the above issues in master/pacific.

**Review by dzafman — APPROVED** [review]
Looks ok

**Pull request merged**